### PR TITLE
[INT-1200] Refactor Explore View

### PR DESCRIPTION
**Pull request by emmajsadams**
This PR refactors the explore view into two DropdownCombobox components. One for project/repo and the other for branches. When a project/repo is selected master is mounted by default. Full details on the behavior and implementation choice [can be found on the story](https://pachyderm.atlassian.net/browse/INT-1200) and shown in the video. The short relevant summary is that Downshift is a mature combobox library that provides the core functionality without any unnecessary styles or unused code. 

Tests were added for the DropdownCombobox component and the mounted api status logic. Existing tests were updated and maintained.

Since I had to go through many of the cypress tests I made a few changes to those. I was able to remove the `wait(3000)` in the BeforeEach loop for the tests, reduce the error timeout (which I can probably bring even lower in a follow up PR), generally use TestIDs a bit more, and finally consolidate some of the setup commands.

To tackle in a follow up PR 
* Adding support for choosing different commits https://github.com/pachyderm/pachyderm/pull/9796#discussion_r1514688752
* Simplify the backend https://pachyderm.atlassian.net/browse/INT-1234

https://github.com/pachyderm/pachyderm/assets/401518/45d863ae-821a-409b-bc10-b801948dd984

repos dropdown

commits dropdown

### CODE REVIEW

**Comment by smalyala** [comment]
1) Think I might have found a bug. If a repo doesn't have branch `master`, the extension crashes.
2) How do I get commits to appear in the dropdown? I'm only seeing branches. 
3) Nit: in the video, it seems like commits are sorted lexicographically. Maybe it's valuable to keep them in recency order, as returned in the RPC? This might make less sense though if there are commits associated with multiple branches, as then the displayed commit list would be an interleaving of commits from separate branches. Just a thought, feel free to ignore.

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    selectItem,
+  } = useCombobox({
+    items: inputItems,
+    initialIsOpen: initialSelectedItem ? false : true,
+    initialSelectedItem: initialSelectedItem,
+    onInputValueChange: ({inputValue}) => {
+      setInputItems(
+        inputValue
+          ? items.filter((item) =>
+              item.toLowerCase().startsWith(inputValue.toLowerCase()),
```
Comment (smalyala) [inline jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx:38]: Is it possible to support matches in the middle of the string? It might be slower, but could be useful if there's many project/repo combos in the displayed list and a user starts searching with the repo name. Having to include the project name in every search might be cumbersome.

File: jupyter-extension/src/plugins/mount/api.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+export const mount = async (
+  updateData: (response: ListMountsResponse) => void,
+  projectRepo: string,
+  branch = 'master',
+): Promise<void> => {
+  const [project, repo] = projectRepo.split('/');
+
+  await unmountAll();
```
Comment (kevyang) [inline jupyter-extension/src/plugins/mount/api.ts:22]: i think there's an opportunity to simplify the backend as well as remove this logic. with the assumption that only one branch is mounted at a time, things get a lot simpler. we can discuss doing this in a later revision if need be.

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx (hunk 1 above)
Comment (emmajsadams) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx:38]: Good idea, let me look into this!

**Comment by emmajsadams** [comment]
> Think I might have found a bug. If a repo doesn't have branch master, the extension crashes.
> How do I get commits to appear in the dropdown? I'm only seeing branches.
> Nit: in the video, it seems like commits are sorted lexicographically. Maybe it's valuable to keep them in recency order, as returned in the RPC? This might make less sense though if there are commits associated with multiple branches, as then the displayed commit list would be an interleaving of commits from separate branches. Just a thought, feel free to ignore.

I'll look into these points further.

> I might be a bit behind on the state of the file browser pagination, but it would be awesome if we could integrate that here in the future!
> Related to point 5, if I have more files than the vertical length of the file browser, I'm unable to scroll down with the mouse. I'm able to scroll down by single-clicking a file and scrolling down with the arrow keys. However, if I scroll down below the last visible file in the file browser, I then see this:

Ahh I think the changes to using `Panel` and some other style changes actually just broke the FileBrowser code. It should be all integrated here, but it seems to be overflowing past the screen now. I'll do some testing with more files to figure out what's going on.

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css
[the remark this answers is not in the document]
Comment (emmajsadams) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css:1]: I'll move them to `utils/components` for the sake of this pr, I do wonder if maybe it makes sense to move components to `src/components` in the long run. Seems strange to have components nested under utils I would normally consider them their own separate thing (a reusable UI element)

File: jupyter-extension/src/plugins/mount/api.ts (hunk 2 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/api.ts:22]: Good point, let me take a look at what would be involved in doing that. If not in this PR definitely a follow up right after.

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css
[the remark this answers is not in the document]
Comment (kevyang) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css:1]: i have no strong opinions here, we can go with moving them to `src/components` if that seems more intuitive to you. i have very little frontend dev experience

File: jupyter-extension/src/plugins/mount/api.ts (hunk 2 above)
Comment (kevyang) [reply jupyter-extension/src/plugins/mount/api.ts:22]: happy to help walk you through the file manager code in the backend when you are ready to do this!

File: jupyter-extension/src/plugins/mount/mount.tsx [hunk 3]
```diff
@@ -138,24 +138,18 @@ export class MountPlugin implements IMountPlugin {
       'pfs',
     );
 
-    this._exploreScreen = new SplitPanel({orientation: 'vertical'});
```
Comment (bbonenfant) [inline jupyter-extension/src/plugins/mount/mount.tsx:141]: I think the SplitPanel was what made the the PFS browser viewable. You might want to consider going back to if you can't get the Panel to work

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css
[the remark this answers is not in the document]
Comment (emmajsadams) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.css:1]: Sounds good regardless of which way we go on this, I think you are right that it should be tackled in another PR.

**Comment by emmajsadams** [comment]
> How do I get commits to appear in the dropdown? I'm only seeing branches.
@smalyala so as far as I can see we actually didn't support switching to specific commits prior to these changes? I did look into what it would take to add that support and it seems like a decent amount of work. I think we should probably tackle this with the overall refactor of the backend @kevyang suggested in a follow up PR. This PR would just keep the current behavior of only letting repo/project/branch be selected.

What makes this a little confusing is how the Jupyter extension creates branches named `commit__`

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx (hunk 1 above)
Comment (emmajsadams) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx]: I pulled in the https://github.com/kentcdodds/match-sorter library to provide a much fuzzier search so I think that should work out here.

**Comment by bbonenfant** [comment]
> What makes this a little confusing is how the Jupyter extension creates branches named `commit__`. That's what you were seeing in the video! 
> Screenshot 2024-03-07 at 4 04 28 PM

To give a little context for why this is the way it is: We have that context (or companion) repo where we upload the "environment" for the user's pipeline. We need to be able to identify which commit of the context repo is paired with the version of the pipeline the user just published. At the time we made this (I think this is still true, but might not be) you couldn't mark a specific commit as an input to a pipeline, so we decided to use single commit branches to accomplish this.
As you can see here, the naming convention was somewhat arbitrary, we just wanted to use the commit hash for uniqueness: https://github.com/pachyderm/pachyderm/blob/master/jupyter-extension/jupyterlab_pachyderm/pps_client.py#L217-L219

So, tldr it's a bit of a hack. Ideally we want to make this context repo an established "type" of repo within pachyderm so that it's probably associated with the pipeline and cleaned up. Unfortunately we haven't gotten around to that yet

**Comment by emmajsadams** [comment]
Ok I believe all the feedback has been addressed! The only follow up here is asking nbredesign about a "sort by recently selected feature", simplifying the backend based on this new workflow, and adding a commit DropdownCombobox. I will create follow up stories for each of those. 

@smalyala just to provide brief responses on each

> Think I might have found a bug. If a repo doesn't have branch master, the extension crashes.

This now selects `master` falling back to the first branch from an sorted list of branches. 

> How do I get commits to appear in the dropdown? I'm only seeing branches.
> Nit: in the video, it seems like commits are sorted lexicographically. Maybe it's valuable to keep them in recency order, as returned in the RPC? This might make less sense though if there are commits associated with multiple branches, as then the displayed commit list would be an interleaving of commits from separate branches. Just a thought, feel free to ignore.

Commits aren't able to be selected since the functionality was not there before. Follow up story can tackle this though.

> If I'm inside a folder and change the repo, the file browser doesn't change. It's confusing to see one repo's set of files but a different repo in the dropdown bar. If I click /pfs in the file browser, it backs out and then I see the right repo's folder.

Changing the repo now updates the FileBrowser. Unmounting changes the directory to root with no folders visible. Mounting a repo changes the directory to the newly mounted folder which allows the user to skip the step of clicking on that folder also.

> I might be a bit behind on the state of the file browser pagination, but it would be awesome if we could integrate that here in the future!
> Related to point 5, if I have more files than the vertical length of the file browser, I'm unable to scroll down with the mouse. I'm able to scroll down by single-clicking a file and scrolling down with the arrow keys. However, if I scroll down below the last visible file in the file browser, I then see this:

This was just a regression I introduced by switching to Panel, this is now fixed by switching back to SplitPanel and adjusting the height in css.

**Comment by smalyala** [comment]
Awesome! Thanks for addressing all those bugs. I definitely mistook those branch names as commits.

One remnant I noticed with regards to branches is if a repo has zero branches, the branch dropdown box disappears. It might be more informative if instead there's a message displayed along the lines of "No branches". Also, the server logs 500 errors. I think if we add a check to the Python layer if branches don't exist, we should be able to avoid that error.
image

File: jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx (hunk 1 above)
Comment (smalyala) [reply jupyter-extension/src/components/DropdownCombobox/DropdownCombobox.tsx]: Awesome, this works great!

File: jupyter-extension/src/plugins/mount/mountDrive.ts [hunk 4]
```diff
@@ -118,6 +118,16 @@ export class MountDrive implements Contents.IDrive {
       return this._getCachedContent();
     }
 
+    // Reset cache
```
Comment (bbonenfant) [inline jupyter-extension/src/plugins/mount/mountDrive.ts:121]: Out of curiosity, why was this moved? It might be relevant to add some context to this comment.

**Review by bbonenfant — APPROVED** [review]
Left a couple of comments. One question about the UI though, when I open up the jupyter server in my browser the project/repo drop down is already there. Not sure if this is a browser issue or if this is intended. Seems to me like it would make more sense for it to drop down when your text cursor is in the box.
image

File: jupyter-extension/src/plugins/mount/mount.tsx [hunk 5]
```diff
@@ -51,7 +51,7 @@ export class MountPlugin implements IMountPlugin {
   private _fullPageError: ReactWidget;
   private _configScreen: ReactWidget;
   private _pipelineScreen: ReactWidget;
-  private _exploreScreen: SplitPanel;
+  private _exploreScreen: Panel;
```
[the remark this answers is not in the document]
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mount.tsx:54]: Ah yep good point will revert

File: jupyter-extension/src/plugins/mount/mountDrive.ts (hunk 4 above)
Comment (emmajsadams) [reply jupyter-extension/src/plugins/mount/mountDrive.ts:121]: I'll add some additional context. Previously the cache was only reset once the user navigated to a new directory that needed the cache not when any change in directory happens. This means if the loading of the files for a directory was still happening when a user unmounted they would get an error shown to them since it would fail to load.

**Comment by emmajsadams** [comment]
> @bbonenfant  Left a couple of comments. One question about the UI though, when I open up the jupyter server in my browser the project/repo drop down is already there. Not sure if this is a browser issue or if this is intended. Seems to me like it would make more sense for it to drop down when your text cursor is in the box.

It's just a setting in the dropdown menu! I can change it to be off by default.

> @smalyala One remnant I noticed with regards to branches is if a repo has zero branches, the branch dropdown box disappears. It might be more informative if instead there's a message displayed along the lines of "No branches". Also, the server logs 500 errors. I think if we add a check to the Python layer if branches don't exist, we should be able to avoid that error.

Good catch I actually did not realize it was possible for a repo to have no branches!

**Review by smalyala — APPROVED** [review]
LGTM pending changes to account for repo with zero branches

**Comment by emmajsadams** [comment]
@smalyala for repos with no branches my thought was maybe we should just filter them out as selectable options? I went ahead and made that change with a new test. Curious if you think that is ok vs maybe allowing them to be selected then showing an error?

**Comment by smalyala** [comment]
If a user selects a repo with no branches, can we prevent the dropdown component from dropping down and instead show a message that says "No branches"? To me, it makes sense to show all repos so users don't wonder why a repo is missing from the list.

**Comment by emmajsadams** [comment]
@smalyala yeah that makes sense I'll make those changes today!

**Comment by emmajsadams** [comment]
@smalyala so regarding fixing the 500 error from the endpoint, as @kevyang already suggested we should probably just simplify the mounting and unmounting endpoints in a follow up story since now the concept of mounting multiple branches is not necessary https://pachyderm.atlassian.net/browse/INT-1234. 

The extension no longer makes the mount request if no branches exist, it shows an error explaining to the user there are no branches then unselects that value. Video here

https://github.com/pachyderm/pachyderm/assets/401518/0a6e8a1b-6fd4-473b-aa19-e2e1c32676f5

**Comment by smalyala** [comment]
Awesome! Looks good!

Re: the 500 error, that makes complete sense.
